File: src/deadlock_build_sync/purchase_categories.py
"""Project the canonical purchase guide into Steam rows and optional choices."""

from __future__ import annotations

from typing import TYPE_CHECKING

from .purchase_types import MAX_CATEGORY_DESCRIPTION_BYTES, GuideCategory

if TYPE_CHECKING:
    from .match_choices import AutomaticBranch
    from .purchase_guidance_types import PurchaseChoice, PurchaseGuidance
    from .purchase_types import GuideItem, PurchaseGuide


def split_guidance(text: str) -> tuple[str, ...]:
    """Split text into bounded rows.

    Returns:
        All input text in UTF-8 bounded rows.

    """
    result: list[str] = []
    while text:
        end = 0
        size = 0
        for character in text:
            width = len(character.encode("utf-8"))
            if size + width > MAX_CATEGORY_DESCRIPTION_BYTES:
                break
            end += 1
            size += width
        if end < len(text):
            boundary = text.rfind(" ", 0, end)
            if boundary > 0:
                end = boundary + 1
        result.append(text[:end])
        text = text[end:]
    return tuple(result)


def choice_instruction(guidance: PurchaseGuidance, card: PurchaseChoice) -> str:
    route = " -> ".join(guidance.names[item] for item in card.route)
    if card.after_step is None:
        return f"{card.purpose.trigger}. Timing unknown. Route: {route}. Catalog cost: {card.catalog_cost:,} souls. Select a checkpoint before purchase."
    path = guidance.default_path.actions
    resume = (
        path[card.after_step].name if card.after_step < len(path) else "core complete"
    )
    cost = (
        f"Extra cost: {card.extra_path_cost:,} souls."
        if card.extra_path_cost is not None
        else "Cost unavailable."
    )
    rebuy = ", ".join(guidance.names[item] for item in card.rebought_components)
    return (
        f"{card.purpose.trigger}. After core step {card.after_step}. Route: {route}. "
        f"{cost} Resume: {resume}."
        + (f" Rebuy: {rebuy}." if rebuy else "")
        + (f" Blocked: {card.blocked_reason}." if card.blocked_reason else "")
        + conditional_instruction(guidance, card)
    )


def _choice_rows(
    guidance: PurchaseGuidance, card: PurchaseChoice, item: GuideItem
) -> list[GuideCategory]:
    decision = next(
        (row for row in guidance.decisions if card.item_id in row.options), None
    )
    kind = (
        decision.kind if decision else "UPGRADE" if len(card.route) > 1 else "OPTIONAL"
    )
    instruction = choice_instruction(guidance, card)

    return [
        GuideCategory(
            f"{kind} | {card.name}" + (f" ({index + 1})" if index else ""),
            (item,),
            part,
            optional=True,
        )
        for index, part in enumerate(split_guidance(instruction))
    ]


def purchase_categories(guide: PurchaseGuide) -> tuple[GuideCategory, ...]:
    guidance = guide.purchase_guidance
    if guidance is None:
        raise ValueError("Purchase categories require a canonical guide")
    items = {item.item_id: item for group in guide.tiers.values() for item in group}
    core = guide.core_purchase_items or guide.core_items
    result: list[GuideCategory] = []
    for checkpoint in range(len(core) + 1):
        if checkpoint:
            step = guidance.default_path.actions[checkpoint - 1]
            result.append(
                GuideCategory(
                    f"CORE {checkpoint}",
                    (core[checkpoint - 1],),
                    f"AUTO QUEUE | Buy step {checkpoint}. Cost: {step.incremental_cost:,} souls. Total: {step.cumulative_cost:,} souls.",
                )
            )
        for card in guidance.choices:
            if card.after_step == checkpoint:
                result.extend(_choice_rows(guidance, card, items[card.item_id]))
        result.extend(_conditional_rows(guidance, checkpoint, items))
    for card in guidance.choices:
        if card.after_step is None:
            result.extend(_choice_rows(guidance, card, items[card.item_id]))
    result.extend(_core_alternatives(guide))
    result.extend(_pool_rows(guide))
    queued = tuple(
        item.item_id
        for category in result
        if not category.optional
        for item in category.items
    )
    if queued != tuple(step.item_id for step in guidance.default_path.actions):
        raise ValueError("Steam Queue differs from the canonical component path")
    return tuple(result)


def category_records(categories: tuple[GuideCategory, ...]) -> list[dict[str, object]]:
    return [
        {
            "name": category.name,
            "optional": category.optional,
            "description": category.description,
            "width": category.width,
            "height": category.height,
            "items": [
                {
                    "item_id": item.item_id,
                    "item": item.name,
                    "annotation": item.annotation,
                    "required_flex_slots": item.required_flex_slots,
                    "sell_priority": item.sell_priority,
                    "imbue_target_ability_id": item.imbue_target_ability_id,
                }
                for item in category.items
            ],
        }
        for category in categories
    ]


def _core_alternatives(guide: PurchaseGuide) -> list[GuideCategory]:
    result: list[GuideCategory] = []
    for alternative in guide.core_alternatives:
        item = next(
            value
            for value in guide.optional_core_items
            if value.item_id == alternative.item_id
        )
        instruction = f"{alternative.when} {alternative.swap}. {alternative.why} {alternative.skip}"
        result.extend(
            GuideCategory("OPTIONAL CORE", (item,), part, optional=True)
            for part in split_guidance(instruction)
        )
    return result


def conditional_instruction(guidance: PurchaseGuidance, card: PurchaseChoice) -> str:
    return "".join(
        _branch_instruction(guidance, branch)
        for branch in guidance.automatic_branches
        if branch.item_id == card.item_id
    )


def _branch_instruction(guidance: PurchaseGuidance, branch: AutomaticBranch) -> str:
    condition = (
        f"wealth is {branch.value} (personal net worth / lobby mean; behind <0.90, ahead >1.10)"
        if branch.condition == "relative_wealth"
        else f"enemy hero {branch.value} is present"
        if branch.condition == "enemy_hero"
        else f"enemy owns {guidance.names.get(int(branch.value), str(branch.value))}"
    )
    instruction = f" IF {condition}, choose this item after step {branch.after_step}."
    if branch.substituted_core:
        instruction += f" Replace {guidance.names[branch.comparator_item_id]}; use its separately validated core path."
    return instruction


def _conditional_rows(
    guidance: PurchaseGuidance, checkpoint: int, items: dict[int, GuideItem]
) -> list[GuideCategory]:
    result = []
    cards = {card.item_id: card for card in guidance.choices}
    for branch in guidance.automatic_branches:
        if branch.after_step != checkpoint:
            continue
        card = cards[branch.item_id]
        plan = branch.default_plan
        if plan is None:
            raise ValueError("Automatic branch has no canonical purchase plan")
        route = " -> ".join(guidance.names[item] for item in card.route)
        steps = guidance.default_path.actions
        resume_at = checkpoint + 1 if branch.substituted_core else checkpoint
        resume = steps[resume_at].name if resume_at < len(steps) else "core complete"
        extra = plan.remaining_cost - guidance.default_path.remaining_cost
        instruction = (
            _branch_instruction(guidance, branch)
            + f" Route: {route}. Extra cost: {extra:,} souls. Resume: {resume}."
        )
        result.extend(
            GuideCategory(
                "OPTIONAL | CONDITIONAL", (items[card.item_id],), part, optional=True
            )
            for part in split_guidance(instruction)
        )
    return result


def _pool_rows(guide: PurchaseGuide) -> list[GuideCategory]:
    return [
        GuideCategory(
            f"ITEM POOL | TIER {tier}",
            guide.tiers.get(tier, ()),
            "Optional items. These tiers are not a purchase order."
            if guide.tiers.get(tier)
            else "No supported options are available.",
            optional=True,
        )
        for tier in range(1, 5)
    ]

File: crates/deadlock-guides/src/purchase_categories.rs
use std::collections::BTreeMap;

use deadlock_data::{Error, Result};

use crate::guide_category::GuideCategory;
use crate::guide_item::{GuideItem, format_integer};
use crate::purchase_guidance_types::{PurchaseChoice, PurchaseGuidance};
use crate::purchase_guide::PurchaseGuide;
use crate::purchase_instructions::{
    format_choice_instruction, format_conditional_route, split_guidance,
};

/// # Errors
/// Returns an error when the category queue, item references, or automatic plans disagree with the canonical guide.
pub fn build_purchase_categories(guide: &PurchaseGuide) -> Result<Vec<GuideCategory>> {
    let guidance = guide
        .purchase_guidance
        .as_ref()
        .ok_or_else(|| Error::new("Purchase categories require a canonical guide"))?;
    let items = guide
        .tiers
        .values()
        .flatten()
        .map(|item| (item.item_id, item))
        .collect::<BTreeMap<_, _>>();
    let core = guide.core_path_items();
    if core.len() != guidance.default_path.actions.len() {
        return Err(Error::new(
            "Purchase category count differs from its canonical path",
        ));
    }
    let mut result = Vec::new();
    for checkpoint in 0..=core.len() {
        if checkpoint > 0 {
            let step = &guidance.default_path.actions[checkpoint - 1];
            let description = format!(
                "AUTO QUEUE | Buy step {checkpoint}. Cost: {} souls. Total: {} souls.",
                format_integer(i128::from(step.incremental_cost)),
                format_integer(i128::from(step.cumulative_cost))
            );
            result.push(GuideCategory::new(
                format!("CORE {checkpoint}"),
                vec![core[checkpoint - 1].clone()],
                description,
                false,
                false,
            )?);
        }
        for card in guidance
            .choices
            .iter()
            .filter(|card| card.after_step == Some(checkpoint))
        {
            result.extend(choice_categories(
                guidance,
                card,
                require_item(&items, card.item_id)?,
            )?);
        }
        result.extend(conditional_categories(guidance, checkpoint, &items)?);
    }
    for card in guidance
        .choices
        .iter()
        .filter(|card| card.after_step.is_none())
    {
        result.extend(choice_categories(
            guidance,
            card,
            require_item(&items, card.item_id)?,
        )?);
    }
    result.extend(core_alternative_categories(guide)?);
    result.extend(item_pool_categories(guide)?);
    let queued = result
        .iter()
        .filter(|category| !category.optional)
        .flat_map(|category| &category.items)
        .map(|item| item.item_id);
    if !queued.eq(guidance
        .default_path
        .actions
        .iter()
        .map(|step| step.item_id))
    {
        return Err(Error::new(
            "Steam Queue differs from the canonical component path",
        ));
    }
    Ok(result)
}

fn require_item<'items>(
    items: &BTreeMap<u64, &'items GuideItem>,
    id: u64,
) -> Result<&'items GuideItem> {
    items
        .get(&id)
        .copied()
        .ok_or_else(|| Error::new(format!("Purchase category references unknown item {id}")))
}

fn choice_categories(
    guidance: &PurchaseGuidance,
    card: &PurchaseChoice,
    item: &GuideItem,
) -> Result<Vec<GuideCategory>> {
    let decision = guidance
        .decisions
        .iter()
        .find(|decision| decision.options.contains(&card.item_id));
    let kind = decision.map_or_else(
        || {
            if card.route.len() > 1 {
                "UPGRADE"
            } else {
                "OPTIONAL"
            }
        },
        |decision| decision.kind.as_str(),
    );
    split_guidance(&format_choice_instruction(guidance, card)?)
        .into_iter()
        .enumerate()
        .map(|(index, part)| {
            let suffix = if index == 0 {
                String::new()
            } else {
                format!(" ({})", index + 1)
            };
            GuideCategory::new(
                format!("{kind} | {}{suffix}", card.name),
                vec![item.clone()],
                part,
                true,
                false,
            )
        })
        .collect()
}

fn conditional_categories(
    guidance: &PurchaseGuidance,
    checkpoint: usize,
    items: &BTreeMap<u64, &GuideItem>,
) -> Result<Vec<GuideCategory>> {
    let cards = guidance
        .choices
        .iter()
        .map(|card| (card.item_id, card))
        .collect::<BTreeMap<_, _>>();
    let mut result = Vec::new();
    for branch in guidance
        .automatic_branches
        .iter()
        .filter(|branch| branch.after_step == checkpoint)
    {
        let card = cards
            .get(&branch.item_id)
            .ok_or_else(|| Error::new("Automatic branch has no purchase choice"))?;
        let item = require_item(items, card.item_id)?;
        for part in split_guidance(&format_conditional_route(guidance, branch, card)?) {
            result.push(GuideCategory::new(
                "OPTIONAL | CONDITIONAL".into(),
                vec![item.clone()],
                part,
                true,
                false,
            )?);
        }
    }
    Ok(result)
}

fn core_alternative_categories(guide: &PurchaseGuide) -> Result<Vec<GuideCategory>> {
    let items = guide
        .optional_core_items
        .iter()
        .map(|item| (item.item_id, item))
        .collect::<BTreeMap<_, _>>();
    let mut result = Vec::new();
    for alternative in &guide.core_alternatives {
        let row = alternative.content();
        let item = require_item(&items, row.item_id)?;
        let instruction = format!("{} {}. {} {}", row.when, row.swap, row.why, row.skip);
        for part in split_guidance(&instruction) {
            result.push(GuideCategory::new(
                "OPTIONAL CORE".into(),
                vec![item.clone()],
                part,
                true,
                false,
            )?);
        }
    }
    Ok(result)
}

fn item_pool_categories(guide: &PurchaseGuide) -> Result<Vec<GuideCategory>> {
    (1..=4)
        .map(|tier| {
            let items = guide.tiers.get(&tier).cloned().unwrap_or_default();
            let description = if items.is_empty() {
                "No supported options are available."
            } else {
                "Optional items. These tiers are not a purchase order."
            };
            GuideCategory::new(
                format!("ITEM POOL | TIER {tier}"),
                items,
                description.into(),
                true,
                false,
            )
        })
        .collect()
}
